### ISSUE TITLE
NodeClaims stranded even after NodePool deletion

### ISSUE THREAD

**Issue by k24dizzle** [aws/karpenter-provider-aws#6905]
### Description

**Observed Behavior**:
- Delete NodePool
- Some NodeClaims (associated with this NodePool) are marked for deletion, but stuck in this state for hours:
```
Name:         infra-t9kvm
Namespace:    
...
API Version:  karpenter.sh/v1
Kind:         NodeClaim
Metadata:
  Creation Timestamp:             2024-08-31T01:00:37Z
  Deletion Grace Period Seconds:  0
  Deletion Timestamp:             2024-08-31T01:09:30Z
  Finalizers:
    karpenter.sh/termination
  Generate Name:  infra-
  Generation:     2
  Owner References:
    API Version:           karpenter.sh/v1
    Block Owner Deletion:  true
    Kind:                  NodePool
    Name:                  infra
    UID:                   ...
  Resource Version:        27623
  UID:                     ...
...
Events:
  Type    Reason             Age                 From       Message
  ----    ------             ----                ----       -------
  Normal  Unconsolidatable   59m                 karpenter  Can't replace with a cheaper node
  Normal  DisruptionBlocked  48s (x29 over 56m)  karpenter  Cannot disrupt NodeClaim: state node is marked for deletion
```

- The karpenter controller doesn't seem to have any logs that signal it is attempting to terminate the node/delete the node claim
<img width="1719" alt="Screenshot 2024-08-30 at 7 06 08 PM" src="https://github.com/user-attachments/assets/b8bd164b-65d7-4c69-83fb-b91c1c92aa21">

- The node itself doesn't have any PDB's that would prevent deletion.
```
Name:               ...
...
Non-terminated Pods:          (3 in total)
  Namespace                   Name                CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                ------------  ----------  ---------------  -------------  ---
  kube-system                 aws-node-w6v4x      50m (5%)      0 (0%)      0 (0%)           0 (0%)         3h4m
  kube-system                 kube-proxy-lz2bw    100m (10%)    0 (0%)      0 (0%)           0 (0%)         3h4m
  loki                        promtail-xgpdr      0 (0%)        0 (0%)      0 (0%)           0 (0%)         3h4m
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests    Limits
  --------           --------    ------
  cpu                150m (15%)  0 (0%)
  memory             0 (0%)      0 (0%)
  ephemeral-storage  0 (0%)      0 (0%)
  hugepages-1Gi      0 (0%)      0 (0%)
  hugepages-2Mi      0 (0%)      0 (0%)
Events:
  Type    Reason             Age                    From       Message
  ----    ------             ----                   ----       -------
  Normal  DisruptionBlocked  3m47s (x86 over 173m)  karpenter  Cannot disrupt Node: state node is marked for deletion
```

**Expected Behavior**:
- Once the node pool is deleted, given that there are no pending pods, attempt to delete all the old nodes associated with the node pool

**Reproduction Steps** (Please include YAML):

**Versions**:
- Chart Version: v1.0.0
- Kubernetes Version (`kubectl version`): 1.29.7

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by engedaam** [comment]
Are the underlaying nodes deleted?

**Comment by k24dizzle** [comment]
> Are the underlaying nodes deleted?

`omni eks-node-viewer --resources cpu --extra-labels karpenter.sh/nodepool --node-sort karpenter.sh/nodepool`

<img width="1552" alt="Screenshot 2024-09-02 at 11 06 04 AM" src="https://github.com/user-attachments/assets/88d44721-9587-45db-87e7-8ee9f827ac19">


The nodes still exist, but stuck in Deleting.

**Comment by engedaam** [comment]
Are there any pods that maybe stuck deleting on those nodes?

**Comment by hamishforbes** [comment]
I'm running into this same issue upgrading to Karpenter 1.0.
I haven't deleted my nodepool but it has been changed to the the v1 custom resource.

So it looks like karpenter struggling because the owner reference on the nodeclaim is for  the v1beta versions of the nodepool?

```
> k get nodeclaims -o custom-columns='APIVER:.apiVersion,NAME:.metadata.name,OWNER_API_VER:.metadata.ownerReferences[0].apiVersion,OWNERKIND:.metadata.ownerReferences[0].kind'
APIVER            NAME            OWNER_API_VER          OWNERKIND
karpenter.sh/v1   default-6nd8t   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-7bxsr   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-9chj5   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-ct54l   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-d6xv7   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-d98d8   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-d9kpt   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-gpjpg   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-j5fdd   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-j6wxw   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-jr5qf   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-l2fml   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-lk8bd   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-m2r7j   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-m5vdt   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-m7mtb   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-mk6r6   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-mnndr   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-mz62l   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-pkdl4   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-ptfm8   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-s9df7   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-ttzk8   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   default-wfpbj   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-wwr7z   karpenter.sh/v1        NodePool
karpenter.sh/v1   default-xphxg   karpenter.sh/v1beta1   NodePool
```

I've got nodes that are empty and should be terminated dangling around, the ec2 instances still running.

```
Non-terminated Pods:          (5 in total)
  Namespace                   Name                                         CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                         ------------  ----------  ---------------  -------------  ---
  kube-system                 aws-node-termination-handler-zcnfs           10m (0%)      100m (2%)   64Mi (0%)        64Mi (0%)      3d15h
  kube-system                 aws-node-zlkl9                               50m (1%)      0 (0%)      0 (0%)           0 (0%)         3d15h
  kube-system                 istio-cni-node-jljp6                         50m (1%)      200m (5%)   200Mi (2%)       400Mi (5%)     3d15h
  kube-system                 kube-proxy-pfzxq                             100m (2%)     0 (0%)      0 (0%)           0 (0%)         3d15h
  prometheus                  prometheus-prometheus-node-exporter-mtfdf    50m (1%)      150m (3%)   32Mi (0%)        32Mi (0%)      3d15h

Events:
  Type     Reason                 Age                    From       Message
  ----     ------                 ----                   ----       -------
  Normal   Unconsolidatable       30m (x321 over 3d15h)  karpenter  SpotToSpotConsolidation is disabled, can't replace a spot node with a spot node
  Normal   DisruptionBlocked      26m                    karpenter  Cannot disrupt Node: not all pods would schedule, knative-eventing/kafka-broker-receiver-746fb7d66f-679g6 => would schedule against uninitialized nodeclaim/default-mk6r6
  Normal   DisruptionTerminating  22m                    karpenter  Disrupting Node: Underutilized/Delete
  Warning  FailedDraining         22m                    karpenter  Failed to drain node, 15 pods are waiting to be evicted
  Normal   DisruptionBlocked      14m (x5 over 22m)      karpenter  Cannot disrupt Node: state node is marked for deletion
  Normal   DisruptionBlocked      2m27s (x6 over 12m)    karpenter  Cannot disrupt Node: state node is marked for deletion
  Normal   DisruptionBlocked      92s                    karpenter  Cannot disrupt Node: state node is marked for deletion
  Normal   DisruptionBlocked      69s                    karpenter  Cannot disrupt Node: state node is marked for deletion
```

The major problem this is causing me is that these dangling nodes often still have PVs mounted, which is preventing those pods being re-scheduled on new nodes (the old multi-attach EBS controller problem, that i'm upgrading to 1.0 to try and fix...)

Manually terminating the EC2 instance does eventually cause everything to clean up.
As does manually updating the apiVersion of the nodepool in the owner reference on the nodeclaim

**Comment by engedaam** [comment]
@hamishforbes are you seeing the empty nodes being deleted in the logs by karpenter?

**Comment by hamishforbes** [comment]
No, that's the problem. 
Any nodeclaim where the ownerRef is for the old v1beta1 nodepool does not get deleted.
Nodes provisioned after the 1.0 upgrade with an ownerRef for the v1 nodepool are fine.

Yes it says that 20 minutes ago there were 15 pods draining. But as you can see there are only daemonset pods left on that node *now*.
If I fix the nodeclaim ownerRef then Karpenter immediately terminates the EC2 instance and cleans up.

**Comment by k24dizzle** [comment]
> Are there any pods that maybe stuck deleting on those nodes?

No, just some running daemonsets.

<img width="1069" alt="Screenshot 2024-09-02 at 10 48 14 PM" src="https://github.com/user-attachments/assets/408cbc77-9b50-495d-ab70-8ac7f828597a">

I'm experiencing this in clusters where only Karpenter 1.0 exists, so don't think its related to upgrading:
```
% kubectl get nodeclaims -o custom-columns='APIVER:.apiVersion,NAME:.metadata.name,OWNER_API_VER:.metadata.ownerReferences[0].apiVersion,OWNERKIND:.metadata.ownerReferences[0].kind'
APIVER            NAME          OWNER_API_VER     OWNERKIND
karpenter.sh/v1   infra-4swmq   karpenter.sh/v1   NodePool
karpenter.sh/v1   infra-gp9d4   karpenter.sh/v1   NodePool
karpenter.sh/v1   infra-vxcpn   karpenter.sh/v1   NodePool
```

**Comment by k24dizzle** [comment]
I think it has something to do with the `karpenter.sh/termination` finalizer for the node claim, is there a way I can manually trigger the finalizer to run again (even after the node pool is deleted?). It isn't too clear to me what the finalizer is doing/if it is rerunning and retrying. I don't see any signal in the logs that it is continuing to run, which makes me think it is stuck.

**Comment by engedaam** [comment]
This issue seems like a duplicate of https://github.com/kubernetes-sigs/karpenter/issues/1578. Could we track the investigation on that issue? It would make it better to keep track

**Comment by k24dizzle** [comment]
I think it's slightly different. I've experienced this issue:
- without upgrading to v1 of karpenter
- where the underlying node does not get deleted

**Comment by engedaam** [comment]
@k24dizzle What version of karpenter are you running?

**Comment by k24dizzle** [comment]
v1.0.0

**Comment by aquam8** [comment]
I have the same problem as the author.

I was running v0.37.2, with `webhook.enabled`. I upgraded to 1.0.1 (for CRD and app), updated IAM policy but kept my manifests for nodepool and ec2nc as v1beta1. All was well at that stage.

The next step was about updating my manifests for nodepool and ec2nc to reference the `v1` CRD so i can leverage the new 
 budget's reasons. But as soon as i updated the manifest to `v1` and applied the changes i have encountered issues.

Karpenter logs:
```
{"level":"ERROR","time":"2024-09-09T06:49:17.087Z","logger":"controller","message":"failed listing instance types for mixed-1","commit":"62a726c","controller":"disruption","namespace":"","name":"","reconcileID":"583102f1-95d3-48f1-99b0-0a76cb430d69","error":"resolving node class, ec2nodeclasses.karpenter.k8s.aws \"mx51-eks\" is terminating, treating as not found
"}
{"level":"ERROR","time":"2024-09-09T06:49:18.369Z","logger":"controller","message":"nodePool not ready","commit":"62a726c","controller":"provisioner","namespace":"","name":"","reconcileID":"9bceb5be-935b-4d2d-b587-8154a3b8e17e","NodePool":{"name":"mixed-1"}}
{"level":"INFO","time":"2024-09-09T06:49:18.369Z","logger":"controller","message":"no nodepools found","commit":"62a726c","controller":"provisioner","namespace":"","name":"","reconcileID":"9bceb5be-935b-4d2d-b587-8154a3b8e17e"}
```

The nodepool fails with `Failed resolving NodeClass`.

The nodeclass fails with `WaitingOnNodeClaimTermination  -  Waiting on NodeClaim termination for mixed-1-q25pf, mixed-1-dv8cp`

```
k get nodeclaims -o custom-columns='APIVER:.apiVersion,NAME:.metadata.name,OWNER_API_VER:.metadata.ownerReferences[0].apiVersion,OWNERKIND:.metadata.ownerReferences[0].kind'
APIVER            NAME            OWNER_API_VER          OWNERKIND
karpenter.sh/v1   mixed-1-dv8cp   karpenter.sh/v1beta1   NodePool
karpenter.sh/v1   mixed-1-q25pf   karpenter.sh/v1beta1   NodePool
```

Recovery from there is painful and hit-and-miss I can't get a new node to register until i kill all existing nodes, or remove the finalizer `karpenter.sh/termination` on the nodeclaims/nodes. Sometimes i have had to re-add the ec2nc for everything to get going again. But of course this is highly disruptive and not suitable for PROD upgrade.

I appreciate any assistance on how to address that last leg of the upgrade.

**Comment by sergii-auctane** [comment]
> This issue seems like a duplicate of [kubernetes-sigs/karpenter#1578](https://github.com/kubernetes-sigs/karpenter/issues/1578). Could we track the investigation on that issue? It would make it better to keep track

It's nothing like that issue. I'm using version 1.0.6 and having tons of nodes suck in 
```
  Warning  FailedDraining     7m30s (x4116 over 6d3h)  karpenter  Failed to drain node, 9 pods are waiting to be evicted
  Normal   DisruptionBlocked  2m16s (x7379 over 10d)   karpenter  Cannot disrupt Node: state node is marked for deletion
```
And those 9 pods are DaemonSets. I observe this issue with non-default node pools only. I run consolidation nightly and the default nodepool consolidates, but not another one.

**Comment by FranciscoCross** [comment]
Good afternoon, I have the same problem. I currently delete nodepools as part of nightly shutdown to save costs, but I am experiencing this problem: there is always one active node left after nodepool deletion.

**Comment by aviralharsh** [comment]
"Any updates on this? We're encountering the same issue. Despite no pods running on the node, Karpenter is unable to delete it. 
In the node events, we see the message: `Cannot disrupt NodeClaim: state node is marked for deletion`."

**Comment by ryan-hashemi90** [comment]
We are experiencing a similar issue. After a fresh installation of Karpenter version `1.0.3`, we’ve observed that nodes are hosting only daemonsets and remain stuck in the deleting state. As @sergii-auctane pointed out, this behavior appears to be limited to our non-default node pools (those with taints + labels).

**Comment by aquam8** [comment]
Although I'm not certain that your issue Ryan is related, i will document how i proceeding to upgrade from v0.37.x to v1.0.x.

The actual upgrade of the Karpenter helm chart from v0.37.x to v1.0.x is actually fine for me, the webhook conversion in v0.37. did it job well, and prepared everything for a v1.0 migration. No issue at that stage.
The problem appears when the NodePool & EC2NC v1beta managed by Terraform were then updated to v1. That forces a resource replacement, and the deletion of the existing resources triggered the cascade deletion (and infinite blocking) of the NodeClaims, Nodes etc..  which is fair enough as i don't think that Karpenter is expecting a replacement of these resources at this stage.

So to get around that problem, i simply avoid an in-place update of these resources, and instead provisioned new NodePool and EC2NC on `v1` API Version with a different name (and i included `Weight: 10` to the new NodePool to instruct Karpenter to give priority to this new nodepool when considered. After these were deployed everything continued working as before on the old v1beta converted to v1 resources, no direct changes. But then i triggered a node rotation (by editing the EC2NC userdata - you can use any other way like such as AMI update..), Karpenter diligently created new NodeClaim based on the new v1 NodePool/EC2NC, drained the existing nodes and rescheduled everything onto the new nodepool/nodes. No downtime, no error.

At that stage, i changed my Terraform to delete the old NodePool/Ec2NC resources that had the v1beta version, that got deployed to clean up, no noticeable changes, everything got cleaned up properly. I'm now prepared to a migration to v1.1 where the v1beta resources are deprecated, and i don't expect trouble at this stage 🤞 .

So i strongly recommend you do a rolling upgrade instead of in place upgrade if you care about downtime impact, and want to upgrade to v1.x
Thank you to all the project contributors because Karpenter rocks!

**Comment by rohan-changejar** [comment]
Hello, we are facing this issue on production and this is leading related nodes and pods going into terminating state indefinitely, post which a manual cleanup is required to sort it out.
Have tried all the solutions mentioned above.
Is there anything else we can do to sort this out?
Also is there any permanent fix for this issue being planned
